### ISSUE TITLE
Cannot override OneOf method names to avoid collision with imported proto

### ISSUE THREAD

**Issue by tomkerkhove** [protocolbuffers/protobuf#18063]
**What version of protobuf and what language are you using?**
Version: [Grpc.AspNetCore v2.65.0 which uses Google.Protobuf (>= 3.27.0)](https://www.nuget.org/packages/Grpc.AspNetCore#dependencies-body-tab)
Language: C#

**What operating system (Linux, Windows, ...) and version?**
Windows

**What runtime / compiler are you using (e.g., python version or gcc version)**
Unknown

**What did you do?**
1. Import [Envoy's external processor proto](https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/ext_proc/v3/external_processor.proto)
2. Compilation fails

**What did you expect to see**

Import and successful compilation

**What did you see instead?**

Compilation fails because

However, [BodyMutation has a clear_body property](https://github.com/envoyproxy/envoy/blob/cbfea5b91657c09021df178c5d7a8495fd7f70fe/api/envoy/service/ext_proc/v3/external_processor.proto#L367-L374) which collides with generated `ClearBody` method which is generated as per https://github.com/protocolbuffers/protobuf/blob/b915e9f44e9d979617d7738f7ddd7dd02ccd7c76/csharp/src/Google.Protobuf/Reflection/OneofDescriptor.cs#L184-L189 without option to change naming.

As far as I know, there is no way to influence the name of the `ClearXXX` method nor the name of the .NET property based on https://protobuf.dev/reference/csharp/csharp-generated/, is that correct?

**Comment by mgravell** [comment]
The code [here](https://github.com/protocolbuffers/protobuf/blob/b915e9f44e9d979617d7738f7ddd7dd02ccd7c76/src/google/protobuf/compiler/csharp/csharp_message.cc#L242-L246) suggests that you're right and there is indeed no way to influence the naming here (as an aside, I wonder if it should actually be called `Reset*` rather than `Clear*`, to follow the `System.ComponentModel` convention, that is actively used by many components - but that ship has probably long sailed!). Likewise, the member-name fudging here is [very minimal](https://github.com/protocolbuffers/protobuf/blob/b915e9f44e9d979617d7738f7ddd7dd02ccd7c76/src/google/protobuf/compiler/csharp/csharp_message.cc#L213-L214).

An unfortunate collision of names! Thinking aloud:

- it would be nice if there was some fallback mechanism to override names, a language-specific override extension (I use something similar in some of my code, but it doesn't apply here)
- perhaps `protoc` should emit warnings for collisions with all inbuilt/known 1st-party language conventions
- you could *ask* whether envoy could consider renaming the field in a breaking change (I doubt you'll win this one); there exists `json_name` to avoid *too much* data compat issues (the binary format only uses the numbers) - although even then, there could be unexpected consequences
- you could perhaps risk applying the above change (rename, using `json_name` for the old name) *locally*, see how it goes - with a note not to revert blindly in .proto updates

**Comment by tomkerkhove** [comment]
Thanks for confirming! I doubt Envoy will want to change the field as it's going to break things indeed.

I like the suggestion of providing a fallback or at least making the naming convention configurable given it's used during generation time anyway.

`json_name` did not work for me as it was not honored, but that may be my sloppy testing.

**Comment by googleberg** [comment]
Unfortunately, there is no way to influence these names outside of changing the field name. Setting the `json_name` option will not change the accessors.

@jskeet any ideas?

**Comment by jskeet** [comment]
I think introducing a language-specific annotation to allow the generated property name to be overridden is probably the only long-term solution. I agree we could warn about this - but we can't easily "escape" the collision in the same way that we do for a few other known collisions (e.g. a field called `types`) because we don't know when the collision will be introduced... the same proto should generate the same code, regardless of the order in which fields are introduced. (Introducing a `clear_body` field shouldn't change an existing `ClearBody()` method, for example.)

@googleberg Do any languages have the ability to override property names like this? Note that there will be runtime library aspects that would need changing too, I suspect, but it would be doable.

**Comment by esrauchg** [comment]
Unfortunately the general case of this issue is just how Protobuf works; it is possible to create a set of names in the same scope which don't work in any one language.

We have best-effort adhoc logic in some of the language's code generators to check for and mangle some subset of cases where some collisions would happen. We don't have a way to override names only in one language today, but its definitely on our radar to add in a future Edition. I believe we also just (very unfortunately) don't really do warnings for some cases where we can reasonably detect something is a risky message but will mostly work in general.

The relevant message linked here amounts to this (the oneof is actually not relevant part of the definition for this context)
```
message BodyMutation {
   optional bytes body = 1;
   optional bool clear_body = 2;
}
```

Unfortunately naming fields this way will have bad behavior many languages, so I'm not sure its the best example of one where a per-language override would be a good solution.

Under C++ it will kinda-accidentally successfully compile, only because the clear function of the first field will be marked non-const and the getter of the second field is const, and C++ allows const/non-const overloads. But, I'm pretty sure the C++ overload behavior prefers const where possible, so if someone writes the statement`bodymutation.clear_body();` against the message that'll actually call the getter of the second field (= do nothing) rather than the clearer of the first field (= what anyone writing that statement would have intended).

Basically; Protobuf still may want to figure out how to warn-or-mitigate such cases better, but for this concrete example I think its worth opening an issue with Envoy (pointing back at this discussion) that they really might want to consider doing the hard effort to rename the field because using this combination of names is error-prone even in languages where it works (which is a breaking change, and not so easy depending on how/where the message is used).
